File: src/badezimmer/mdns.py
import logging
import re

import asyncudp
from zeroconf import (
    IPVersion,
    Zeroconf,
)

from badezimmer import (
    DeviceCategory,
    DeviceKind,
    MDNSQueryRequest,
    MDNSQueryResponse,
    MDNSQuestion,
    MDNSType,
    MDNSRecord,
    MDNSPointerRecord,
    MDNSARecord,
    MDNSSRVRecord,
    MDNSTextRecord,
    MDNS,
)
from badezimmer.logger import setup_logger
from badezimmer.tcp import (
    prepare_protobuf_request,
    get_protobuf_data,
    get_all_ips_strings_for_adapters,
)
import asyncio
import random
import time
import threading

setup_logger()
logger = logging.getLogger(__name__)
zeroconf = Zeroconf(ip_version=IPVersion.V4Only)

_MULTICAST_IP = "224.0.0.251"
MULTICAST_PORT = 5369
RANDOM_SEED = 42069
DEFAULT_TTL = 4500
SERVICE_DISCOVERY_TYPE = "_services._dns-sd._udp.local"


class NonUniqueNameException(Exception):
    """Exception when the name is already registered."""


def current_time_millis() -> float:
    return time.monotonic() * 1000


class EntryRecord:
    def __init__(self, record: MDNSRecord):
        self.record = record
        self.expires_at = record.ttl + current_time_millis()

    def expired(self) -> bool:
        return self.expires_at < current_time_millis()


class MDNSServiceInfo:
    def __init__(
        self,
        name: str,
        type: str,
        port: int,
        kind: DeviceKind,
        category: DeviceCategory,
        properties: dict[str, str],
        ttl: int = DEFAULT_TTL,
        allow_name_change: bool = True,
    ):
        self.name = name
        self.type = type
        self.category = category
        self.port = port
        self.properties = properties
        self.kind = kind
        self.ttl = ttl
        self.allow_name_change = allow_name_change
        self.ips = get_all_ips_strings_for_adapters()

    def to_records(self) -> list[MDNSRecord]:
        records = []
        ptr_record = MDNSRecord(
            name=self.type,
            ttl=self.ttl,
            cache_flush=False,
            ptr_record=MDNSPointerRecord(name=self.name, domain_name=self.name),
        )
        records.append(ptr_record)

        domain_name = generate_domain_name(self.type, self.name)

        # Add IPs as A records
        for ip in self.ips:
            record = MDNSRecord(
                name=domain_name,
                ttl=self.ttl,
                cache_flush=True,
                a_record=MDNSARecord(name=domain_name, address=ip),
            )
            records.append(record)

        domain_name_parts = domain_name.split(".")
        srv_record = MDNSRecord(
            name=domain_name,
            ttl=self.ttl,
            cache_flush=True,
            srv_record=MDNSSRVRecord(
                name=domain_name_parts[-1],
                protocol=domain_name_parts[-2],
                service=domain_name_parts[-3],
                instance=domain_name_parts[-4],
                port=self.port,
                target=domain_name,
            ),
        )
        records.append(srv_record)

        txt_record_entries: dict[str, str] = {}

        txt_record_entries["kind"] = DeviceKind.Name(self.kind.numerator)
        txt_record_entries["category"] = DeviceCategory.Name(self.category.numerator)
        for key, value in self.properties.items():
            txt_record_entries[key] = value

        txt_record = MDNSRecord(
            name=domain_name,
            ttl=self.ttl,
            cache_flush=True,
            txt_record=MDNSTextRecord(
                name=domain_name,
                entries=txt_record_entries,
            ),
        )
        records.append(txt_record)
        return records


class BadezimmerMDNS:
    def __init__(
        self,
        interval_between_tiebreaking_in_millis: float = 100.0,
        tiebreaking_attempts: int = 3,
        query_timeout_in_millis: int = 200,
        tiebreaking_max_drift_in_millis: float = 25.0,
        random_seed: int = RANDOM_SEED,
    ):
        self.interval_between_tiebreaking_in_millis = (
            interval_between_tiebreaking_in_millis
        )
        self.tiebreaking_attempts = tiebreaking_attempts
        self.query_timeout_in_millis = query_timeout_in_millis
        self.tiebreaking_max_drift_in_millis = tiebreaking_max_drift_in_millis
        self.random = random.Random(random_seed)
        self.cache: dict[str, list[EntryRecord]] = {}
        self.services: dict[str, MDNSServiceInfo] = {}
        self.loop: asyncio.AbstractEventLoop | None = None
        self._loop_thread: threading.Thread | None = None

    async def register_service(self, info: MDNSServiceInfo):
        logger.debug(
            "Registering service with on Badezimmer Network...",
            extra={
                "service_name": info.name,
                "service_type": info.type,
                "port": info.port,
                "properties": info.properties,
            },
        )

        logger.debug(
            "Sleeping before registering service to avoid thundering herd",
            extra={"info": info},
        )
        await asyncio.sleep(self.random.randint(150, 250) / 1000)
        await self.__tiebreaker(info)
        await self.__add_service(info)
        await self.__broadcast_service(info)

    async def __tiebreaker(self, info: MDNSServiceInfo) -> None:
        logger.debug("Tiebreaking...", extra={"type": info.type})

        current_attempt = 0
        next_instance_number = 2
        current_instance_name = info.name
        now = next_time = current_time_millis()

        while current_attempt < self.tiebreaking_attempts:
            logger.info("Tiebreaking attempt", extra={"attempt": current_attempt + 1})

            while await self.__check_if_service_is_defined(
                entry_name=info.type, instance_name=current_instance_name
            ):
                if not info.allow_name_change:
                    logger.exception(
                        "Service is already defined checking if allow_name_change"
                    )
                    raise NonUniqueNameException

                current_instance_name = f"{info.name}-{next_instance_number}"
                next_instance_number += 1

                # Reset attempt counter
                current_attempt = 0
                next_time = now

            logger.debug("Service is defined")

            if now < next_time:
                await asyncio.sleep(next_time - now)
                now = current_time_millis()
                continue

            # Do another query to check if the service is defined until the max attempts is reached
            await self.__query_service(info.type, self.query_timeout_in_millis)

            current_attempt += 1
            next_time += (self.interval_between_tiebreaking_in_millis / 1000) + (
                self.random.random() * self.tiebreaking_max_drift_in_millis / 1000
            )

        info.name = current_instance_name

        logger.debug(
            "Tiebreaking resolved with success",
            extra={"type": info.type, "service_name": info.name},
        )

    async def __add_service(self, info: MDNSServiceInfo) -> None:
        self.services[info.name] = info

    async def __broadcast_service(self, info: MDNSServiceInfo) -> None:
        records = info.to_records()
        ptr_record = records[0]
        additional_records = records[1:]

        response = MDNSQueryResponse(
            answer=ptr_record, additional_records=additional_records
        )

        packet = MDNS(transaction_id=0x0000, query_response=response)

        raw_bytes = prepare_protobuf_request(packet)
        sock = await asyncudp.create_socket(remote_addr=(_MULTICAST_IP, MULTICAST_PORT))
        sock.sendto(raw_bytes)

    async def __query_service(
        self, type_: str, query_timeout_in_millis: int
    ) -> MDNSQueryResponse | None:
        question = MDNSQuestion(name=type_, type=MDNSType.MDNS_PTR)
        query = MDNSQueryRequest(questions=[question])

        packet = MDNS(transaction_id=0x0000, query_request=query)

        raw_bytes = prepare_protobuf_request(packet)

        try:
            logger.debug("Sending query request to multicast address")

            sock = await asyncudp.create_socket(
                remote_addr=(_MULTICAST_IP, MULTICAST_PORT)
            )
            sock.sendto(raw_bytes)

            try:
                async with asyncio.timeout(query_timeout_in_millis / 1000):
                    resp_raw_bytes = await sock.recvfrom()
                proto_bytes = get_protobuf_data(resp_raw_bytes)
            except TimeoutError:
                logger.error("Timeout while waiting for response on send query")
                return None

            packet = MDNS()
            packet.ParseFromString(proto_bytes)

            if packet.WhichOneof("data") != "query_response":
                raise ValueError("Invalid packet received")

            response = packet.query_response

            # Add answers to cache
            self.__add_entry_record(type_, response.answer)
            [
                self.__add_entry_record(type_, record)
                for record in response.additional_records
            ]

        except Exception as e:
            logger.error(f"Error sending query request to multicast address: {e}")
            raise

        return None

    async def __check_if_service_is_defined(
        self, entry_name: str, instance_name: str
    ) -> bool:
        entry_records = self.cache.get(entry_name)

        if entry_records is None:
            return False

        for entry_record in reversed(entry_records):
            if (
                entry_record.record.WhichOneof("record") != "ptr_record"
                or entry_record.expired()
            ):
                continue

            ptr_record = entry_record.record.ptr_record

            domain_name = generate_domain_name(entry_name, instance_name)
            if ptr_record.domain_name == domain_name:
                return True

        return False

    def __add_entry_record(self, entry_name: str, record: MDNSRecord) -> None:
        self.cache.setdefault(entry_name, []).append(EntryRecord(record))


def generate_domain_name(entry: str, instance_name: str) -> str:
    return f"{instance_name}.{entry}"

File: go-water-leak/mdns.go
package main

import (
	"context"
	"encoding/binary"
	"fmt"
	"log"
	"math/rand"
	"net"
	"sync"
	"syscall"
	"time"

	"github.com/talDoFlemis/badezimmer/go-water-leak/badezimmer"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

const (
	MulticastIP   = "224.0.0.251"
	MulticastPort = 5369
	DefaultTTL    = 4500
	ServiceDiscoveryType = "_services._dns-sd._udp.local"
	
	// SO_REUSEPORT for Linux
	SO_REUSEPORT = 15
)

type MDNSServiceInfo struct {
	Name       string
	Type       string
	Port       int32
	Kind       badezimmer.DeviceKind
	Category   badezimmer.DeviceCategory
	Protocol   badezimmer.TransportProtocol
	Properties map[string]string
	Addresses  []string
	TTL        int32
}

type BadezimmerMDNS struct {
	conn              *net.UDPConn
	registeredServices map[string]*MDNSServiceInfo // key: domain_name
	sentPackets       [][]byte
	sentPacketsMu     sync.Mutex
	ctx               context.Context
	cancel            context.CancelFunc
	wg                sync.WaitGroup
}

func NewBadezimmerMDNS() *BadezimmerMDNS {
	ctx, cancel := context.WithCancel(context.Background())
	return &BadezimmerMDNS{
		registeredServices: make(map[string]*MDNSServiceInfo),
		sentPackets:        make([][]byte, 0, 50),
		ctx:                ctx,
		cancel:             cancel,
	}
}

func (m *BadezimmerMDNS) Start() error {
	addr := &net.UDPAddr{
		IP:   net.ParseIP("0.0.0.0"),
		Port: MulticastPort,
	}

	// Create a listening connection with SO_REUSEPORT to allow multiple processes
	lc := net.ListenConfig{
		Control: func(network, address string, c syscall.RawConn) error {
			var opErr error
			err := c.Control(func(fd uintptr) {
				// Enable SO_REUSEADDR
				opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
				if opErr != nil {
					return
				}
				// Enable SO_REUSEPORT to allow multiple processes to bind to the same port
				opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, SO_REUSEPORT, 1)
			})
			if err != nil {
				return err
			}
			return opErr
		},
	}

	packetConn, err := lc.ListenPacket(context.Background(), "udp4", addr.String())
	if err != nil {
		return fmt.Errorf("failed to listen UDP: %w", err)
	}

	conn, ok := packetConn.(*net.UDPConn)
	if !ok {
		return fmt.Errorf("failed to cast to UDPConn")
	}

	m.conn = conn

	// Join multicast group
	multicastIP := net.ParseIP(MulticastIP)
	
	err = conn.SetReadBuffer(65536)
	if err != nil {
		return fmt.Errorf("failed to set read buffer: %w", err)
	}

	// Get socket file descriptor and set multicast options
	file, err := conn.File()
	if err != nil {
		return fmt.Errorf("failed to get socket file: %w", err)
	}
	defer file.Close()

	fd := int(file.Fd())
	
	// Join multicast group using IP_ADD_MEMBERSHIP
	mreq := &syscall.IPMreq{
		Multiaddr: [4]byte{multicastIP[0], multicastIP[1], multicastIP[2], multicastIP[3]},
		Interface: [4]byte{0, 0, 0, 0}, // Use default interface
	}
	
	if err := syscall.SetsockoptIPMreq(fd, syscall.IPPROTO_IP, syscall.IP_ADD_MEMBERSHIP, mreq); err != nil {
		log.Printf("Warning: failed to join multicast group: %v", err)
	} else {
		log.Printf("Joined multicast group %s", MulticastIP)
	}

	log.Printf("BadezimmerMDNS listening on %s:%d", MulticastIP, MulticastPort)

	// Start receive loop
	m.wg.Add(1)
	go m.recvLoop()

	// Start renovation loop
	m.wg.Add(1)
	go m.renovateLoop()

	return nil
}

func (m *BadezimmerMDNS) Close() error {
	m.cancel()
	
	// Send goodbye packets for all registered services
	for _, info := range m.registeredServices {
		goodbyeInfo := *info
		goodbyeInfo.TTL = 0
		m.broadcastService(&goodbyeInfo)
		log.Printf("Sent goodbye packet for service: %s", info.Name)
	}

	if m.conn != nil {
		m.conn.Close()
	}
	
	m.wg.Wait()
	return nil
}

func (m *BadezimmerMDNS) RegisterService(info *MDNSServiceInfo) error {
	log.Printf("Registering service: %s on port %d", info.Name, info.Port)
	
	// Add random delay
	time.Sleep(time.Duration(150+rand.Intn(100)) * time.Millisecond)
	
	domainName := generateDomainName(info.Type, info.Name)
	m.registeredServices[domainName] = info
	
	// Broadcast service
	return m.broadcastService(info)
}

func (m *BadezimmerMDNS) UnregisterService(info *MDNSServiceInfo) error {
	log.Printf("Unregistering service: %s", info.Name)
	
	domainName := generateDomainName(info.Type, info.Name)
	delete(m.registeredServices, domainName)
	
	// Send goodbye packet
	goodbyeInfo := *info
	goodbyeInfo.TTL = 0
	return m.broadcastService(&goodbyeInfo)
}

func (m *BadezimmerMDNS) UpdateService(info *MDNSServiceInfo) error {
	log.Printf("Updating service: %s", info.Name)
	
	domainName := generateDomainName(info.Type, info.Name)
	m.registeredServices[domainName] = info
	
	return m.broadcastService(info)
}

func (m *BadezimmerMDNS) recvLoop() {
	defer m.wg.Done()
	
	buffer := make([]byte, 65536)
	for {
		select {
		case <-m.ctx.Done():
			return
		default:
		}
		
		m.conn.SetReadDeadline(time.Now().Add(1 * time.Second))
		n, addr, err := m.conn.ReadFromUDP(buffer)
		if err != nil {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				continue
			}
			log.Printf("Error reading from UDP: %v", err)
			continue
		}
		
		data := buffer[:n]
		
		// Skip our own packets
		if m.isSentPacket(data) {
			continue
		}
		
		log.Printf("Received packet from %s (%d bytes)", addr.IP, n)
		m.handlePacket(data, addr)
	}
}

func (m *BadezimmerMDNS) renovateLoop() {
	defer m.wg.Done()
	
	// Renovate at 75% of TTL
	renovationInterval := time.Duration(float64(DefaultTTL) * 0.75) * time.Second
	ticker := time.NewTicker(renovationInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-m.ctx.Done():
			return
		case <-ticker.C:
			count := 0
			for _, info := range m.registeredServices {
				if err := m.broadcastService(info); err != nil {
					log.Printf("Error renovating service %s: %v", info.Name, err)
				} else {
					count++
				}
			}
			if count > 0 {
				log.Printf("TTL renovation cycle completed (%d services)", count)
			}
		}
	}
}

func (m *BadezimmerMDNS) handlePacket(data []byte, addr *net.UDPAddr) {
	protoBytes, err := getProtobufData(data)
	if err != nil {
		log.Printf("Error extracting protobuf data: %v", err)
		return
	}
	
	packet := &badezimmer.MDNS{}
	if err := proto.Unmarshal(protoBytes, packet); err != nil {
		log.Printf("Error unmarshaling MDNS packet: %v", err)
		return
	}
	
	switch packet.GetData().(type) {
	case *badezimmer.MDNS_QueryRequest:
		m.handleQuery(packet.GetQueryRequest(), addr)
	case *badezimmer.MDNS_QueryResponse:
		// We could handle responses here if needed
		log.Printf("Received query response from %s", addr.IP)
	}
}

func (m *BadezimmerMDNS) handleQuery(query *badezimmer.MDNSQueryRequest, addr *net.UDPAddr) {
	var ptrRecords []*badezimmer.MDNSRecord
	var additionalRecords []*badezimmer.MDNSRecord
	
	for _, question := range query.Questions {
		if question.Name == ServiceDiscoveryType {
			// Respond with all our registered services
			for _, info := range m.registeredServices {
				records := infoToRecords(info)
				if len(records) > 0 {
					ptrRecords = append(ptrRecords, records[0])
					additionalRecords = append(additionalRecords, records[1:]...)
				}
			}
		} else {
			// Check if this question matches any of our registered services
			for _, info := range m.registeredServices {
				if info.Type == question.Name {
					records := infoToRecords(info)
					if len(records) > 0 {
						ptrRecords = append(ptrRecords, records[0])
						additionalRecords = append(additionalRecords, records[1:]...)
					}
				}
			}
		}
	}
	
	if len(ptrRecords) > 0 {
		response := &badezimmer.MDNSQueryResponse{
			Answers:           ptrRecords,
			AdditionalRecords: additionalRecords,
		}
		m.sendResponse(response)
	}
}

func (m *BadezimmerMDNS) broadcastService(info *MDNSServiceInfo) error {
	records := infoToRecords(info)
	if len(records) == 0 {
		return fmt.Errorf("no records generated for service")
	}
	
	response := &badezimmer.MDNSQueryResponse{
		Answers:           []*badezimmer.MDNSRecord{records[0]},
		AdditionalRecords: records[1:],
	}
	
	return m.sendResponse(response)
}

func (m *BadezimmerMDNS) sendResponse(response *badezimmer.MDNSQueryResponse) error {
	packet := &badezimmer.MDNS{
		TransactionId: rand.Uint32(),
		Timestamp:     timestamppb.Now(),
		Data:          &badezimmer.MDNS_QueryResponse{QueryResponse: response},
	}
	
	return m.sendPacket(packet)
}

func (m *BadezimmerMDNS) sendPacket(packet *badezimmer.MDNS) error {
	rawBytes, err := prepareProtobufRequest(packet)
	if err != nil {
		return fmt.Errorf("failed to prepare packet: %w", err)
	}
	
	m.addSentPacket(rawBytes)
	
	addr := &net.UDPAddr{
		IP:   net.ParseIP(MulticastIP),
		Port: MulticastPort,
	}
	
	_, err = m.conn.WriteToUDP(rawBytes, addr)
	if err != nil {
		return fmt.Errorf("failed to send packet: %w", err)
	}
	
	log.Printf("Sent packet (%d bytes, txid: %d)", len(rawBytes), packet.TransactionId)
	return nil
}

func (m *BadezimmerMDNS) addSentPacket(data []byte) {
	m.sentPacketsMu.Lock()
	defer m.sentPacketsMu.Unlock()
	
	// Keep last 50 packets
	if len(m.sentPackets) >= 50 {
		m.sentPackets = m.sentPackets[1:]
	}
	m.sentPackets = append(m.sentPackets, data)
}

func (m *BadezimmerMDNS) isSentPacket(data []byte) bool {
	m.sentPacketsMu.Lock()
	defer m.sentPacketsMu.Unlock()
	
	for _, sent := range m.sentPackets {
		if bytesEqual(sent, data) {
			return true
		}
	}
	return false
}

func prepareProtobufRequest(msg proto.Message) ([]byte, error) {
	serialized, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}
	
	length := uint32(len(serialized))
	lengthPrefix := make([]byte, 4)
	binary.BigEndian.PutUint32(lengthPrefix, length)
	
	return append(lengthPrefix, serialized...), nil
}

func getProtobufData(data []byte) ([]byte, error) {
	if len(data) < 4 {
		return nil, fmt.Errorf("data too short for length prefix")
	}
	
	messageLength := binary.BigEndian.Uint32(data[:4])
	if uint32(len(data)-4) < messageLength {
		return nil, fmt.Errorf("data shorter than expected message length")
	}
	
	return data[4 : 4+messageLength], nil
}

func generateDomainName(serviceType, instanceName string) string {
	return fmt.Sprintf("%s.%s", instanceName, serviceType)
}

func infoToRecords(info *MDNSServiceInfo) []*badezimmer.MDNSRecord {
	var records []*badezimmer.MDNSRecord
	domainName := generateDomainName(info.Type, info.Name)
	
	// 1. PTR Record
	ptrRecord := &badezimmer.MDNSRecord{
		Name:       info.Type,
		Ttl:        info.TTL,
		CacheFlush: false,
		Record: &badezimmer.MDNSRecord_PtrRecord{
			PtrRecord: &badezimmer.MDNSPointerRecord{
				Name:       info.Type,
				DomainName: domainName,
			},
		},
	}
	records = append(records, ptrRecord)
	
	// 2. A Records
	for _, ip := range info.Addresses {
		aRecord := &badezimmer.MDNSRecord{
			Name:       domainName,
			Ttl:        info.TTL,
			CacheFlush: true,
			Record: &badezimmer.MDNSRecord_ARecord{
				ARecord: &badezimmer.MDNSARecord{
					Name:    domainName,
					Address: ip,
				},
			},
		}
		records = append(records, aRecord)
	}
	
	// 3. SRV Record
	service := "_http"
	if parts := splitServiceType(info.Type); len(parts) > 0 {
		service = parts[0]
	}
	
	srvRecord := &badezimmer.MDNSRecord{
		Name:       domainName,
		Ttl:        info.TTL,
		CacheFlush: true,
		Record: &badezimmer.MDNSRecord_SrvRecord{
			SrvRecord: &badezimmer.MDNSSRVRecord{
				Name:     info.Name,
				Protocol: info.Protocol,
				Service:  service,
				Instance: info.Name,
				Port:     info.Port,
				Target:   domainName,
			},
		},
	}
	records = append(records, srvRecord)
	
	// 4. TXT Record
	txtEntries := make(map[string]string)
	txtEntries["kind"] = info.Kind.String()
	txtEntries["category"] = info.Category.String()
	for k, v := range info.Properties {
		txtEntries[k] = v
	}
	
	txtRecord := &badezimmer.MDNSRecord{
		Name:       domainName,
		Ttl:        info.TTL,
		CacheFlush: true,
		Record: &badezimmer.MDNSRecord_TxtRecord{
			TxtRecord: &badezimmer.MDNSTextRecord{
				Name:    domainName,
				Entries: txtEntries,
			},
		},
	}
	records = append(records, txtRecord)
	
	return records
}

func splitServiceType(serviceType string) []string {
	result := []string{}
	for i := 0; i < len(serviceType); i++ {
		if serviceType[i] == '.' {
			result = append(result, serviceType[:i])
			break
		}
	}
	return result
}

func getLocalIPv4Addresses() []string {
	var addresses []string
	
	ifaces, err := net.Interfaces()
	if err != nil {
		return addresses
	}
	
	excludedPrefixes := []string{"127.", "172.17.", "172.18.", "172.19.", "172.20.", "172.21.", "172.22."}
	
	for _, iface := range ifaces {
		addrs, err := iface.Addrs()
		if err != nil {
			continue
		}
		
		for _, addr := range addrs {
			var ip net.IP
			switch v := addr.(type) {
			case *net.IPNet:
				ip = v.IP
			case *net.IPAddr:
				ip = v.IP
			}
			
			if ip == nil || ip.To4() == nil {
				continue
			}
			
			ipStr := ip.String()
			excluded := false
			for _, prefix := range excludedPrefixes {
				if len(ipStr) >= len(prefix) && ipStr[:len(prefix)] == prefix {
					excluded = true
					break
				}
			}
			
			if !excluded {
				addresses = append(addresses, ipStr)
			}
		}
	}
	
	return addresses
}

func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
